### DISCUSSION TITLE
Have Inspector.has_table() expressed as an Expression. Possible ?

### DISCUSSION THREAD

**Discussion by davidandreoletti** [sqlalchemy/sqlalchemy#8465] [Usage Questions]
**Context:**

SQLAlchemy 1.4 recommends to use the [Inspector's has_table](https://docs.sqlalchemy.org/en/14/core/reflection.html#sqlalchemy.engine.reflection.Inspector.has_table) to check whether a table exists. 

The downside is for session, especially with multiple binds where the bind is resolved based on the query's information (eg: Table name, etc) or application specific sharding resolution, ```has_table``` can't be used.

Since each (most?) dialect has ```has_table```, then the goal would for the custom expression to have the exact same scope (in term of database engine types) support as `inspector's has_table` has.

Ideally, one would write some custom expression such as:

```
class HasTable(Executable, ClauseElement):
     def __init__(..., table_name:str=None, schema:str=None):
          self._table_name = table_name
          self._schema = schema


@compiles(HasTable, 'postgres')
def _visit_hastable(element, compiler, **kw):
        // Somehow find the relevant query (or expose via a new method the rendered query available inside PGDialect.has_table)
       return sql_table_present_query
```

Then, execute the query ```table_found = await session.execute(HasTable(table="foobar", schema"test"))```. 

**Q1. Is it a good way to approach this ? 
Q1.1 If not, what's the alternative ?**

**Comment by CaselIT** [comment]
Hi,

Just to understand the problem, I don't think the change from engine.has_table to inspector.has_table has any impact on this issue, since with multiple engines you would still need to run the query multiple times as you would with the inspector.

What you would like to do is to specify a session with multiple binds, then query for has table and that should select to the correct engine?
Not sure how popular is this problem, since if you know how to select the proper engine changes are you know if the table is there or not.

In any case you can already do this with the current architecture, getting a connection from the session with multiple binds, inspecting that connection and running has_table there.
[`get_connection()`](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=session+connection#sqlalchemy.orm.Session.connection) accepts `bind_arguments` so it allows choosing the connection from the proper engine
